File: solutions/lua/isbn-verifier/1/isbn-verifier.lua
return {
  valid = function(isbn)
    if isbn:find("[^-%dX]") or isbn:find(".+X.+") then return false end
    local w, sum = 10, 0
    for ch in isbn:gmatch("[%dX]") do
      local v = (ch == "X" and 10) or ch:byte() - ("0"):byte()
      sum, w = sum + w * v, w - 1
    end
    return w == 0 and sum % 11 == 0
  end
}

File: solutions/ruby/isbn-verifier/1/isbn_verifier.rb
# frozen_string_literal: true

class IsbnVerifier
  def self.valid?(isbn)
    digits = isbn.upcase.gsub(/[^\dX]/,'').to_s
    return false if digits.length != 10
    return false if digits.include?('X') && digits.index('X') != 9
    chksum = 10.downto(1).zip(digits.chars).sum { |n,d| n * (d == 'X' ? 10 : d.to_i) } % 11
    chksum.zero?
  end
end
